// @remix-run/fetch-router/dist/lib/request-context.js
import SuperHeaders from '@remix-run/headers';
import { createSession } from '@remix-run/session';
import { AppStorage } from "./app-storage.js";
import { RequestBodyMethods, } from "./request-methods.js";
/**
 * A context object that contains information about the current request. Every request
 * handler or middleware in the lifecycle of a request receives the same context object.
 */
export class RequestContext {
    constructor(request) {
        this.headers = new SuperHeaders(request.headers);
        this.method = request.method.toUpperCase();
        this.params = {};
        this.request = request;
        this.storage = new AppStorage();
        this.url = new URL(request.url);
    }
    /**
     * A map of files that were uploaded in the request.
     */
    get files() {
        let formData = this.formData;
        if (formData == null) {
            return null;
        }
        let files = new Map();
        for (let [key, value] of formData.entries()) {
            if (value instanceof File) {
                files.set(key, value);
            }
        }
        return files;
    }
    /**
     * Parsed [`FormData`](https://developer.mozilla.org/en-US/docs/Web/API/FormData) from the
     * request body.
     *
     * Note: This is only available for requests with a body (not `GET` or `HEAD`).
     *
     * [MDN Reference](https://developer.mozilla.org/en-US/docs/Web/API/FormData)
     */
    get formData() {
        if (this.#formData == null && RequestBodyMethods.includes(this.method)) {
            this.#formData = new FormData();
        }
        return this.#formData;
    }
    set formData(value) {
        this.#formData = value;
    }
    #formData;
    /**
     * The headers of the request.
     */
    headers;
    /**
     * The request method. This may differ from `request.method` if the request body contained a
     * method override field (e.g. `_method=DELETE`), allowing HTML forms to simulate RESTful API
     * request methods like `PUT` and `DELETE`.
     */
    method;
    /**
     * Params that were parsed from the URL.
     */
    params;
    /**
     * The original request that was dispatched to the router.
     */
    request;
    /**
     * The current session.
     */
    get session() {
        if (this.#session == null) {
            console.warn("Session isn't started yet, so session data won't be saved. Use the session() middleware to start the session.");
            this.#session = createSession();
        }
        return this.#session;
    }
    set session(value) {
        this.#session = value;
    }
    #session;
    /**
     * Whether the session has been started.
     */
    get sessionStarted() {
        return this.#session != null;
    }
    /**
     * Shared application-specific storage.
     */
    storage;
    /**
     * The URL that was matched by the route.
     *
     * Note: This may be different from `request.url` if the request was routed to a sub-router,
     * in which case the sub-router's mount path is stripped from the beginning of the pathname.
     */
    url;
}
