// @remix-run/fetch-router/dist/lib/route-map.js
import { RoutePattern } from '@remix-run/route-pattern';
export class Route {
    method;
    pattern;
    constructor(method, pattern) {
        this.method = method;
        this.pattern = typeof pattern === 'string' ? new RoutePattern(pattern) : pattern;
    }
    href(...args) {
        return this.pattern.href(...args);
    }
    match(url) {
        return this.pattern.match(url);
    }
}
export function createRoutes(baseOrDefs, defs) {
    return typeof baseOrDefs === 'string' || baseOrDefs instanceof RoutePattern
        ? buildRouteMap(typeof baseOrDefs === 'string' ? new RoutePattern(baseOrDefs) : baseOrDefs, defs)
        : buildRouteMap(new RoutePattern('/'), baseOrDefs);
}
function buildRouteMap(base, defs) {
    let routes = {};
    for (let key in defs) {
        let def = defs[key];
        if (def instanceof Route) {
            routes[key] = new Route(def.method, base.join(def.pattern));
        }
        else if (typeof def === 'string' || def instanceof RoutePattern) {
            routes[key] = new Route('ANY', base.join(def));
        }
        else if (typeof def === 'object' && def != null && 'pattern' in def) {
            routes[key] = new Route(def.method ?? 'ANY', base.join(def.pattern));
        }
        else {
            routes[key] = buildRouteMap(base, def);
        }
    }
    return routes;
}
