// tempo.ts/dist/wagmi/Actions/nonce.js
import { Actions } from 'viem/tempo';
/**
 * Gets the nonce for an account and nonce key.
 *
 * @example
 * ```ts
 * import { createConfig, http } from '@wagmi/core'
 * import { tempo } from 'viem/chains'
 * import { Actions } from 'tempo.ts/wagmi'
 *
 * const config = createConfig({
 *   chains: [tempoTestnet],
 *   transports: {
 *     [tempo.id]: http(),
 *   },
 * })
 *
 * const nonce = await Actions.nonce.getNonce(config, {
 *   account: '0x...',
 *   nonceKey: 1n,
 * })
 * ```
 *
 * @param config - Config.
 * @param parameters - Parameters.
 * @returns The nonce value.
 */
export function getNonce(config, parameters) {
    const { chainId, ...rest } = parameters;
    const client = config.getClient({ chainId });
    return Actions.nonce.getNonce(client, rest);
}
(function (getNonce) {
    function queryKey(parameters) {
        return ['getNonce', parameters];
    }
    getNonce.queryKey = queryKey;
    function queryOptions(config, parameters) {
        const { query, ...rest } = parameters;
        return {
            ...query,
            queryKey: queryKey(rest),
            async queryFn({ queryKey }) {
                const [, { account, nonceKey, ...parameters }] = queryKey;
                if (!account)
                    throw new Error('account is required.');
                if (nonceKey === undefined)
                    throw new Error('nonceKey is required.');
                return await getNonce(config, { account, nonceKey, ...parameters });
            },
        };
    }
    getNonce.queryOptions = queryOptions;
})(getNonce || (getNonce = {}));
/**
 * Watches for nonce incremented events.
 *
 * @example
 * ```ts
 * import { createConfig, http } from '@wagmi/core'
 * import { tempo } from 'viem/chains'
 * import { Actions } from 'tempo.ts/wagmi'
 *
 * const config = createConfig({
 *   chains: [tempoTestnet],
 *   transports: {
 *     [tempo.id]: http(),
 *   },
 * })
 *
 * const unwatch = Actions.nonce.watchNonceIncremented(config, {
 *   onNonceIncremented: (args, log) => {
 *     console.log('Nonce incremented:', args)
 *   },
 * })
 * ```
 *
 * @param config - Config.
 * @param parameters - Parameters.
 * @returns A function to unsubscribe from the event.
 */
export function watchNonceIncremented(config, parameters) {
    const { chainId, ...rest } = parameters;
    const client = config.getClient({ chainId });
    return Actions.nonce.watchNonceIncremented(client, rest);
}
